tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/method/FirebaseSubscriptionSetCreatePerformer.scala

```diff
@@ -0,0 +1,82 @@
+    def asMessageSetError: SetError = e match {
+      case e: FirebaseSubscriptionCreationParseException => e.setError
+      case e: ExpireTimeInvalidException => SetError.invalidArguments(SetErrorDescription(e.getMessage), Some(Properties("expires")))
+      case e: DeviceClientIdInvalidException => SetError.invalidArguments(SetErrorDescription(e.getMessage))
+      case e: IllegalArgumentException => SetError.invalidArguments(SetErrorDescription(e.getMessage))
+      case _ => SetError.serverFail(SetErrorDescription(e.getMessage))
+    }
+  }
+
+  case class CreationResults(results: Seq[CreationResult]) {
+    def created: Option[Map[FirebaseSubscriptionCreationId, FirebaseSubscriptionCreationResponse]] =
+      Option(results.flatMap {
+        case result: CreationSuccess => Some((result.clientId, result.response))
+        case _ => None
+      }.toMap)
+        .filter(_.nonEmpty)
+
+    def notCreated: Option[Map[FirebaseSubscriptionCreationId, SetError]] = {
+      Option(results.flatMap {
+        case failure: CreationFailure => Some((failure.clientId, failure.asMessageSetError))
+        case _ => None
+      }.toMap)
+        .filter(_.nonEmpty)
+    }
+  }
+}
+
+class FirebaseSubscriptionSetCreatePerformer @Inject()(val repository: FirebaseSubscriptionRepository,
+                                                       val serializer: FirebaseSubscriptionSerializer) {
+
+  def create(request: FirebaseSubscriptionSetRequest, username: Username): SMono[CreationResults] =
+    SFlux.fromIterable(request.create.getOrElse(Map()))
+      .concatMap {
+        case (clientId, json) => parseCreate(json)
+          .fold(e => SMono.just[CreationResult](CreationFailure(clientId, e)),
+            creationRequest => create(clientId, creationRequest, username))
+      }.collectSeq()
+      .map(CreationResults)
+
+
+  private def parseCreate(json: JsObject): Either[Exception, FirebaseSubscriptionCreationRequest] = for {
+    validJsObject <- FirebaseSubscriptionCreation.validateProperties(json)
+    parsedRequest <- serializer.deserializeFirebaseSubscriptionCreationRequest(validJsObject).asEither
+      .left.map(errors => FirebaseSubscriptionCreationParseException.from(errors))
+    validatedRequest <- parsedRequest.validate
+      .left.map(e => FirebaseSubscriptionCreationParseException(SetError.invalidArguments(SetErrorDescription(e.getMessage))))
+  } yield {
+    validatedRequest
+  }
+
+  private def create(clientId: FirebaseSubscriptionCreationId, request: FirebaseSubscriptionCreationRequest, username: Username): SMono[CreationResult] =
+    SMono.fromPublisher(repository.save(username, request))
+      .map(subscription => CreationSuccess(clientId, FirebaseSubscriptionCreationResponse(subscription.id, showExpires(subscription.expires, request))))
+      .onErrorResume(e => SMono.just[CreationResult](CreationFailure(clientId, e)))
```

Comment:
I would love to see us validate the input token here (only available in https://github.com/linagora/tmail-backend/pull/519 though)

Reply:
I just looked in the firebase library. It hasn't token validation method (client side). 
Should we post a request to FCM to validate the token?

Reply:
We have: https://github.com/linagora/tmail-backend/pull/519#discussion_r1014977239
However, we are discussing whether to use it.

Reply:
> I would love to see us validate the input token here (only available in https://github.com/linagora/tmail-backend/pull/519 though)

Nice remark!

See https://github.com/linagora/tmail-backend/issues/520